lib/utils/proxy/proxy_test.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/lib/utils/socks"
+)
+
+// TestDialProxy tests the dialing mechanism of HTTP and SOCKS proxies.
+func TestDialProxy(t *testing.T) {
+	ctx := context.Background()
+	dest := "remote-ip:3080"
+
+	cases := []struct {
+		proxy      func(chan error, net.Listener)
+		scheme     string
+		assertDial require.ErrorAssertionFunc
+	}{
+		{
+			proxy:      serveSOCKSProxy,
+			scheme:     "socks5",
+			assertDial: require.NoError,
+		},
+		{
+			proxy:      serveHTTPProxy,
+			scheme:     "http",
+			assertDial: require.NoError,
+		},
+		{
+			proxy:      func(errChan chan error, l net.Listener) { close(errChan) },
+			scheme:     "unknown",
+			assertDial: require.Error,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.scheme, func(t *testing.T) {
+			errChan := make(chan error)
+			l, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+
+			t.Cleanup(func() {
+				require.NoError(t, l.Close())
+				err := <-errChan
+				require.NoError(t, err)
+			})
+
+			go tc.proxy(errChan, l)
+
+			proxyURL, err := url.Parse(tc.scheme + "://" + l.Addr().String())
+			require.NoError(t, err)
+
+			conn, err := client.DialProxy(ctx, proxyURL, dest)
+			tc.assertDial(t, err)
+
+			if conn != nil {
+				result := make([]byte, len(dest))
+				_, err = io.ReadFull(conn, result)
+				require.NoError(t, err)
+				require.Equal(t, string(result), dest)
+			}
+		})
+	}
+}
+
+// serveSOCKSProxy starts a limited SOCKS proxy on the supplied listener.
+// It performs the SOCKS5 handshake then writes back the requested remote address.
+func serveSOCKSProxy(errChan chan error, l net.Listener) {
+	defer close(errChan)
+
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				errChan <- trace.Wrap(err)
```

Comment:
This will "deadlock" as any other write to this channel. `errChan` is unbuffered, so it will block until someone reads from it. You only read from that channel on the test cleanup, so if anything happens before that, this will hang until the test timeout and the cleanup method is called.